### DISCUSSION TITLE
Guides for implementing generic math interfaces

### DISCUSSION THREAD

**Discussion by WhiteBlackGoose** [dotnet/csharplang#5158] [Q&A]
Hello. I want to ask for a guide or maybe article on implementing generic math interfaces.

The question is *not* about *how* to implement them. What I don't fully understand is what interfaces and when I want to implement.

For example, there's `INumber<>`, which inherits a ton of interfaces. I have type `Complex`, which can implement most of `INumber`'s methods, but not, for instance, the `Max` method. Should I implement `INumber` and then throw unsupported wherever a method doesn't make sense? Or should I implement all and only interfaces which make 100% sense for `Complex`, but not implement `INumber` itself?

**Comment by FaustVX** [comment]
I think you should implement only working interfaces.
I don't think `throw NotSupportedException` in Release version is a good practice.
And when creating API, only support the minimal interface needed :
```csharp
public T SumAll<T>(params T[] values)
where T : IAdditionOperators<T, T, T>, IAdditionIdentity<T> // don't use INumber<T>
{
    var sum = T.Zero;
    foreach (var value in values)
        sum += value;
    return sum;
}
```

**Comment by tannergooding** [comment]
This is probably a better issue for `dotnet/runtime` or `dotnet/docs` than `csharplang`.

That being said, `INumber<T>` is really meant for representing "scalar" numbers (`a`), that is numbers that are representable by a single value. A `Complex` number is really more like a 2-dimensional number (and so is a type of "vector"), as they represent two parts: a `Real` part and an `Imaginary` part (`a + bi`, where `a` is the real part and `bi` represents the imaginary part, which more concretely is a real part `b` and the imaginary unit `i`).

"Vector" like values come with additional complexities and considerations which can come in terms of representation, access, creation, general use, and more. Its something that has already had some thought in place and which should hopefully have a more concrete representation in .NET 7 (and which types like `System.Numerics.Complex` should implement).
